src/components/slider/draw.js

```diff
@@ -0,0 +1,507 @@
+        findDimensions(gd, sliderOpts);
+    }
+
+    sliderGroups.each(function(sliderOpts) {
+        computeLabelSteps(sliderOpts);
+
+        drawSlider(gd, d3.select(this), sliderOpts);
+
+        makeInputProxy(gd, d3.select(this), sliderOpts);
+
+    });
+};
+
+function makeInputProxy(gd, sliderGroup, sliderOpts) {
+    sliderOpts.inputProxy = gd._fullLayout._paperdiv.selectAll('input.' + constants.inputProxyClass)
+        .data([0]);
+}
+
+// This really only just filters by visibility:
+function makeSliderData(fullLayout) {
+    var contOpts = fullLayout[constants.name],
+        sliderData = [];
+
+    for(var i = 0; i < contOpts.length; i++) {
+        var item = contOpts[i];
+        if(item.visible) sliderData.push(item);
+    }
+
+    return sliderData;
+}
+
+// This is set in the defaults step:
+function keyFunction(opts) {
+    return opts._index;
+}
+
+// Compute the dimensions (mutates sliderOpts):
+function findDimensions(gd, sliderOpts) {
+    var sliderLabels = gd._tester.selectAll('g.' + constants.labelGroupClass)
+        .data(sliderOpts.steps);
+
+    sliderLabels.enter().append('g')
+        .classed(constants.labelGroupClass, true);
+
+    // loop over fake buttons to find width / height
+    var maxLabelWidth = 0;
+    var labelHeight = 0;
+    sliderLabels.each(function(stepOpts) {
+        var labelGroup = d3.select(this);
+
+        var text = drawLabel(labelGroup, {step: stepOpts}, sliderOpts);
+
+        var tWidth = text.node() && Drawing.bBox(text.node()).width;
+
+        // This just overwrites with the last. Which is fine as long as
+        // the bounding box (probably incorrectly) measures the text *on
+        // a single line*:
+        labelHeight = text.node() && Drawing.bBox(text.node()).height;
+
+        maxLabelWidth = Math.max(maxLabelWidth, tWidth);
+    });
+
+    sliderLabels.remove();
+
+    sliderOpts.inputAreaWidth = Math.max(
+        constants.railWidth,
+        constants.gripHeight
+    );
+
+    var graphSize = gd._fullLayout._size;
+    sliderOpts.lx = graphSize.l + graphSize.w * sliderOpts.x;
+    sliderOpts.ly = graphSize.t + graphSize.h * (1 - sliderOpts.y);
+
+    if(sliderOpts.lenmode === 'fraction') {
+        // fraction:
+        sliderOpts.outerLength = Math.round(graphSize.w * sliderOpts.len);
+    } else {
+        // pixels:
+        sliderOpts.outerLength = sliderOpts.len;
+    }
+
+    // Set the length-wise padding so that the grip ends up *on* the end of
+    // the bar when at either extreme
+    sliderOpts.lenPad = Math.round(constants.gripWidth * 0.5);
+
+    // The length of the rail, *excluding* padding on either end:
+    sliderOpts.inputAreaStart = 0;
+    sliderOpts.inputAreaLength = Math.round(sliderOpts.outerLength - sliderOpts.xpad * 2);
+    sliderOpts.railInset = Math.round(Math.max(0, constants.gripWidth - constants.railWidth) * 0.5);
+    sliderOpts.stepInset = Math.round(Math.max(sliderOpts.railInset, constants.gripWidth * 0.5));
+
```

Comment:
The logic in this area needs to be 30% less fancy and 40% more directly controlled by constants.